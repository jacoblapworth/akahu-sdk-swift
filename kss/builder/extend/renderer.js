const { regexDictionary, types } = require('../parser/constants');
const tinycolor = require('tinycolor2');

module.exports = function(handlebars) {
	/**
	 * Extracts the variable properties from the line
	 * returns:
	 *  token name, token value, and optional [ token example ]
	 */
	handlebars.registerHelper('renderTokens', renderTokens);

	/**
	 * Compares the name to see if the container needs to be rendered as inverted.
	 * Similar to the isEqual
	 *
	 * e.g.
	 * {{#isInverted name}}
	 * 		ADD THIS
	 * {{else}} [optional]
	 * 		WILL EXECUTE THIS
	 * {{/isInverted}}
	 */
	handlebars.registerHelper('isInverted', isInverted);
};

const testString = 'To understand what recursion is, you must first understand recursion.';


function renderTokens(variables, block) {
	var newBlock = '';
	variables.split('\n').forEach(line => {
		var test;
		const matches = line.match(/[^:]+/g).map(match => match.trim());

		if (matches[1]) {
			this.token = {};
			this.token.name = matches[0];
			this.token.value = matches[1];
			this.token.title = matches[3];
			this.token.description = matches[4];
			if (matches[2] != undefined) {
				this.token.example = getExample(matches[2], this.token);
			}
			newBlock += block.fn(this);
		}
	});
	return newBlock;
}

function getExample(type, token) {
	switch(type) {
		case types.color:
			const codeTextClass = tinycolor(token.value).isDark() ?
				'ds-colour-code-light'
				: null;
			return `
					<div class="ds-colour-example" style="background-color:${token.value}">
						<code class="${codeTextClass}">${token.value}</code>
						<code class="${codeTextClass}">${token.name}</code>
					</div>`;
		case types.border:
			return `
				<div class="ds-colour-example" style="border:${token.value}">
					<code>${token.name}</code>
				</div>`;
		case types.shadow:
			return `
					<div class="ds-colour-example" style="box-shadow:${token.value}">
						<code>${token.name}</code>
					</div>`;
		case types.fontSize:
			return  `<span class="xui-text-truncated" style="font-size:${token.value};">Aa</span>`;
		case types.fontWeight:
			return `<span style="font-weight:${token.value};">Aa</span>`;
		case types.lineHeight:
			return `<span style="line-height:${token.value}">${testString}</span>`;
		case types.spacing:
			return `<div class="ds-spacing-example" style="height:${token.value};width:${token.value};"></div>`;
		default:
			return '';
	}
}

function isInverted(name, options) {
	return regexDictionary.inverted.test(name) ? options.fn(this) : options.inverse(this);
}
