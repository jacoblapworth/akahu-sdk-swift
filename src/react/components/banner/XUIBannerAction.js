import React from 'react';
import PropTypes from 'prop-types';
import XUIButton from '../../button';
import cn from 'classnames';

export default function XUIBannerAction(props) {
	const className = cn(props.className, 'xui-button-small');
	const buttonQaHook = props.qaHook && `${props.qaHook}--button`;

	return (
		<li className="xui-banner--action" data-automationid={props.qaHook}>
			<XUIButton onClick={props.onClick} href={props.href} isLink={props.isLink} variant="link" className={className} qaHook={buttonQaHook}>{props.children}</XUIButton>
		</li>
	);
}

XUIBannerAction.propTypes = {
	className: PropTypes.string,
	children: PropTypes.node,
	qaHook: PropTypes.string,
	/** Click event handler for the banner action */
	onClick: PropTypes.func,
	/** URL of the link */
	href: PropTypes.string,
	/** Whether or not to render this button using an anchor element */
	isLink: PropTypes.bool
};

XUIBannerAction.defaultProps = {
	isLink: false
};
