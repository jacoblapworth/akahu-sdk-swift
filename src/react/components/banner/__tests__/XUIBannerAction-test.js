import React from 'react';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
import renderer from 'react-test-renderer';
import XUIBannerAction from '../XUIBannerAction';

Enzyme.configure({ adapter: new Adapter() });

describe('<XUIBannerAction />', () => {

    it('should render as a button be default and includes no automation id', () => {
        const test = <XUIBannerAction> Action </XUIBannerAction>;
        
        const href = renderer.create(test);
        expect(href).toMatchSnapshot();

        const jestDom = shallow(test);
        expect(jestDom.childAt(0).props().isLink).toBeFalsy();
        expect(jestDom.props()["data-automationId"]).toBe(undefined);
    });

    it('should render a passed qaHook as an auotmation id', () => {
        const qaHookValue = "banner-action";
        const test = <XUIBannerAction qaHook={qaHookValue}>Action</XUIBannerAction>;
       
        const automationId = renderer.create(test);
        expect(automationId).toMatchSnapshot();

        const jestDom = shallow(test);
        expect(jestDom.props()["data-automationId"]).toEqual(qaHookValue);
    });

    it('should render as a link if an href value is passed and isLink prop is true', () => {
        const test = <XUIBannerAction href="http://xero.com" isLink={true}> Action </XUIBannerAction>;
        
        const href = renderer.create(test);
        expect(href).toMatchSnapshot();

        const jestDom = shallow(test);
        expect(jestDom.childAt(0).props().isLink).toBeTruthy();
    });

    it('should render custom classes on the button DOM node', () => {
        const className = "example-class";
        const test = <XUIBannerAction className={className}> Action </XUIBannerAction>;

        const classes = renderer.create(test);
        expect(classes).toMatchSnapshot();

        const jestDom = shallow(test);
        expect(jestDom.childAt(0).props().className).toContain(className);
    });
});