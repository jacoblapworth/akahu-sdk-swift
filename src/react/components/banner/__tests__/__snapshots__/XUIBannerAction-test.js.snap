// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<XUIBannerAction /> should render a passed qaHook as an auotmation id 1`] = `
<li
  className="xui-banner--action"
  data-automationId="banner-action"
>
  <button
    className="xui-button xui-button-small xui-button-borderless-main"
    data-automationid="banner-action--button"
    disabled={false}
    onClick={[Function]}
    onKeyDown={undefined}
    tabIndex={0}
    type="button"
  >
    Action
  </button>
</li>
`;

exports[`<XUIBannerAction /> should render as a button be default and includes no automation id 1`] = `
<li
  className="xui-banner--action"
  data-automationId={undefined}
>
  <button
    className="xui-button xui-button-small xui-button-borderless-main"
    data-automationid={undefined}
    disabled={false}
    onClick={[Function]}
    onKeyDown={undefined}
    tabIndex={0}
    type="button"
  >
     Action 
  </button>
</li>
`;

exports[`<XUIBannerAction /> should render as a link if an href value is passed and isLink prop is true 1`] = `
<li
  className="xui-banner--action"
  data-automationId={undefined}
>
  <a
    className="xui-button xui-button-small xui-button-borderless-main"
    data-automationid={undefined}
    disabled={false}
    href="http://xero.com"
    onClick={[Function]}
    onKeyDown={undefined}
    onKeyPress={[Function]}
    rel={undefined}
    tabIndex={0}
    target={undefined}
  >
     Action 
  </a>
</li>
`;

exports[`<XUIBannerAction /> should render custom classes on the button DOM node 1`] = `
<li
  className="xui-banner--action"
  data-automationId={undefined}
>
  <button
    className="xui-button example-class xui-button-small xui-button-borderless-main"
    data-automationid={undefined}
    disabled={false}
    onClick={[Function]}
    onKeyDown={undefined}
    tabIndex={0}
    type="button"
  >
     Action 
  </button>
</li>
`;
